### DISCUSSION TITLE
Add use.readr option to Feature Request: ...

### DISCUSSION THREAD

**Discussion by iangow** [duckdb/duckdb#7561] [Ideas]
Many users of `duckdb` surely use `dplyr`/Tidyverse packages. The function `readr::read_csv()` is much more robust and flexible than the base `read.csv()` function (and the former tends to do a much better job of inferring type than the latter). It would be great to have the ability to specify the use of `readr`.

Below I include an indicative version of `duckdb_read_csv()` (using `readr::read_delim()`) in a reprex using data available [here](https://github.com/cathytanimura/sql_book/tree/master/Chapter%206:%20Anomaly%20Detection).

Note that the only change to the function body is the inclusion of 

```r
    if (!use.readr) {
      headers <- lapply(files, utils::read.csv, sep = delim, na.strings = na.strings, 
                        quote = quote, nrows = nrow.check, header = header, ...)
    } else {
      headers <- lapply(files, readr::read_delim, delim = delim, na = na.strings, 
                        quote = quote, n_max = nrow.check, guess_max = nrow.check, col_names = header, 
                        show_col_types = FALSE, ...)
```

in place of 

```r
 headers <- lapply(files, utils::read.csv, sep = delim, na.strings = na.strings, 
                   quote = quote, nrows = nrow.check, header = header, ...)
``` 

## Reprex

``` r
setwd("/Users/iangow/git/sql_book")
library(DBI)
library(dplyr, warn.conflicts = FALSE)
library(readr)

duckdb_read_csv <-
  function (conn, name, files, header = TRUE, na.strings = "", 
            nrow.check = 500, delim = ",", quote = "\"", col.names = NULL, 
            lower.case.names = FALSE, sep = delim, transaction = TRUE, use.readr = FALSE,
            ...) {
    
    if (length(na.strings) > 1) 
      stop("na.strings must be of length 1")
    if (!missing(sep)) 
      delim <- sep
    if (!use.readr) {
      headers <- lapply(files, utils::read.csv, sep = delim, na.strings = na.strings, 
                        quote = quote, nrows = nrow.check, header = header, ...)
    } else {
      headers <- lapply(files, readr::read_delim, delim = delim, na = na.strings, 
                        quote = quote, n_max = nrow.check, guess_max = nrow.check, col_names = header, 
                        show_col_types = FALSE, ...)
    }
    if (length(files) > 1) {
      nn <- sapply(headers, ncol)
      if (!all(nn == nn[1])) 
        stop("Files have different numbers of columns")
      nms <- sapply(headers, names)
      if (!all(nms == nms[, 1])) 
        stop("Files have different variable names")
      types <- sapply(headers, function(df) sapply(df, dbDataType, 
                                                   dbObj = conn))
      if (!all(types == types[, 1])) 
        stop("Files have different variable types")
    }
    if (transaction) {
      dbBegin(conn)
      on.exit(tryCatch(dbRollback(conn), error = function(e) {
      }))
    }
    tablename <- dbQuoteIdentifier(conn, name)
    if (!dbExistsTable(conn, tablename)) {
      if (lower.case.names) 
        names(headers[[1]]) <- tolower(names(headers[[1]]))
      if (!is.null(col.names)) {
        if (lower.case.names) {
          warning("Ignoring lower.case.names parameter as overriding col.names are supplied.")
        }
        col.names <- as.character(col.names)
        if (length(unique(col.names)) != length(names(headers[[1]]))) {
          stop("You supplied ", length(unique(col.names)), 
               " unique column names, but file has ", length(names(headers[[1]])), 
               " columns.")
        }
        names(headers[[1]]) <- col.names
      }
      dbWriteTable(conn, tablename, headers[[1]][FALSE, , drop = FALSE])
    }
    for (i in seq_along(files)) {
      thefile <- dbQuoteString(conn, enc2native(normalizePath(files[i])))
      dbExecute(conn, sprintf("COPY %s FROM %s (DELIMITER %s, QUOTE %s, HEADER %s, NULL %s)", 
                              tablename, thefile, dbQuoteString(conn, delim), dbQuoteString(conn, 
                                                                                            quote), tolower(header), dbQuoteString(conn, 
                                                                                                                                   na.strings[1])))
    }
    out <- dbGetQuery(conn, paste("SELECT COUNT(*) FROM", tablename))[[1]]
    if (transaction) {
      dbCommit(conn)
      on.exit(NULL)
    }
    invisible(out)
  }



files <- list.files("data", pattern = "earthquakes.*csv", full.names = TRUE)
files
#>  [1] "data/earthquakes1.csv"  "data/earthquakes10.csv" "data/earthquakes11.csv"
#>  [4] "data/earthquakes12.csv" "data/earthquakes13.csv" "data/earthquakes14.csv"
#>  [7] "data/earthquakes15.csv" "data/earthquakes2.csv"  "data/earthquakes3.csv" 
#> [10] "data/earthquakes4.csv"  "data/earthquakes5.csv"  "data/earthquakes6.csv" 
#> [13] "data/earthquakes7.csv"  "data/earthquakes8.csv"  "data/earthquakes9.csv"
db <- dbConnect(duckdb::duckdb())
duckdb_read_csv(db, "earthquakes", files, use.readr = TRUE,
                col_types = list(time = col_datetime()))
earthquakes <- tbl(db, "earthquakes")

earthquakes
#> # Source:   table<earthquakes> [?? x 22]
#> # Database: DuckDB 0.7.1 [root@Darwin 22.5.0:R 4.3.0/:memory:]
#>    time                latitude longitude depth   mag magtype   nst   gap
#>    <dttm>                 <dbl>     <dbl> <dbl> <dbl> <chr>   <dbl> <dbl>
#>  1 2010-01-01 00:10:43     61.0     -152. 88.4   1.6  ml         NA    NA
#>  2 2010-01-01 00:16:49     48.2     -122.  2.95  2.2  md         16    79
#>  3 2010-01-01 00:16:54     36.0     -118.  1.10  1.1  ml         19    71
#>  4 2010-01-01 00:29:29     60.1     -141.  0     1.4  ml         NA    NA
#>  5 2010-01-01 00:30:16     33.5     -116.  4.38  0.75 ml         13   210
#>  6 2010-01-01 00:31:40     63.2     -150. 86.7   1    ml         NA    NA
#>  7 2010-01-01 00:34:29     36.0     -118.  1.99  0.46 ml          8   118
#>  8 2010-01-01 00:55:05     53.9     -164. 22.4   2.4  ml         NA    NA
#>  9 2010-01-01 00:57:31     63.3     -151.  6.2   0.6  ml         NA    NA
#> 10 2010-01-01 01:09:53     38.8     -123.  1.60  0.2  md          9    67
#> # ℹ more rows
#> # ℹ 14 more variables: dmin <dbl>, rms <dbl>, net <chr>, id <chr>,
#> #   updated <dttm>, place <chr>, type <chr>, horizontalerror <dbl>,
#> #   deptherror <dbl>, magerror <dbl>, magnst <dbl>, status <chr>,
#> #   locationsource <chr>, magsource <chr>

dbDisconnect(db, shutdown = TRUE)
```

<sup>Created on 2023-05-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by iangow** [comment]
Here are some StackOverflow questions where the function above might be relevant:

 - [Read multiple files ...](https://stackoverflow.com/questions/73628489/read-multiple-files-into-duckdb-in-r-from-csv-with-new-variable-indicating-year)
 - [Duckdb_read_csv struggling with with auto detecting column data types in R](https://stackoverflow.com/questions/74445176/duckdb-read-csv-struggling-with-with-auto-detecting-column-data-types-in-r)
 - [Read multiple files into DuckDB in R ...](https://stackoverflow.com/questions/73628489/read-multiple-files-into-duckdb-in-r-from-csv-with-new-variable-indicating-year)

**Comment by iangow** [comment]
@krlmlr Have I missed anything here?

**Comment by rsund** [comment]
Adding readr as a new dependency (even soft one) may be a problem. It also seems that the R function may not take into account all new features of DuckDB's own CSV-reader (e.g. `union_by_name`).

Btw, have you tried with:
`earthquakes <- dplyr::tbl("read_csv_auto('data/earthquakes*.csv')"`? See [docs](https://duckdb.org/docs/data/csv/overview.html) for parameters and alternative versions.

**Comment by iangow** [comment]
Editing slightly to `earthquakes <- dplyr::tbl(db, "read_csv_auto('data/earthquakes*.csv')")` does indeed work. (In fact it seems shockingly fast.)

It took me a little while to work out how to make the data persist. But this seemed to do the trick:

```r
dbfile <- "data/earthquakes.duckdb"

if (!file.exists(dbfile)) {
  db <- dbConnect(duckdb::duckdb(), dbdir = dbfile, read_only = FALSE)
  earthquakes <- 
    tbl(db, "read_csv_auto('data/earthquakes*.csv')") |>
    compute(name = "earthquakes", temporary = FALSE)

dbDisconnect(db, shutdown = TRUE)
```

I guess it would be ideal to have something in between `read_csv_auto` and `read_csv` whereby one could specify the type for some fields and let `read_csv_auto` guess the rest. Something like this:

```r
 earthquakes <- 
      tbl(db, "read_csv_auto('data/earthquakes*.csv', columns={'time': 'TIMESTAMP'})"
```

**Reply by rsund** [reply]
Would #6756 help if COLUMS-parameter is not (already) working with read_csv_auto and if you coudn't define all necessary parameters to read_csv?

**Reply by iangow** [reply]
> Would #6756 help if `COLUMNS`-parameter is not (already) working with `read_csv_auto` and if you coudn't define all necessary parameters to `read_csv`?

My experience is that functions *like* `read_csv_auto()` (e.g., `readr::read_csv()`) can often guess most columns, but there will be a column or two where it doesn't guess correctly. Being able to inspect the results from `read_csv_auto()` and then run again with specified types for incorrect guesses (e.g., something analogous to `col_types = list(time = "DATETIME")`) facilitates this.

In contrast, I don't think #6756 would support this approach. Data like that in `data/earthquakes*.csv'`—if not handled properly by `read_csv_auto()` with default settings—would need at least `auto_type_candidates=['FLOAT', 'VARCHAR', 'DATETIME']`, but there would then seem to be no way to force `time` to be `DATETIME` instead of `VARCHAR`. Going back to `read_csv()` and specifying *all* column types would be comparatively cumbersome.

Note that `read_csv_auto()` appears to guess `DATETIME` correctly for `data/earthquakes*.csv'` with the default parameters, so perhaps it's not the best test case.

**Reply by Mytherin** [reply]
`read_csv_auto` supports overriding column types of a subset of the columns, see [here](https://duckdb.org/docs/data/csv/tips).

```csv
id	date	str
1	2020-01-01	hello world
```

```sql
D FROM read_csv_auto('test.csv');
┌───────┬────────────┬─────────────┐
│  id   │    date    │     str     │
│ int64 │    date    │   varchar   │
├───────┼────────────┼─────────────┤
│     1 │ 2020-01-01 │ hello world │
└───────┴────────────┴─────────────┘
D FROM read_csv_auto('test.csv', types={'date': 'VARCHAR'});
┌───────┬────────────┬─────────────┐
│  id   │    date    │     str     │
│ int64 │  varchar   │   varchar   │
├───────┼────────────┼─────────────┤
│     1 │ 2020-01-01 │ hello world │
└───────┴────────────┴─────────────┘
```

**Reply by iangow** [reply]
> `read_csv_auto` supports overriding column types of a subset of the columns, see [here](https://duckdb.org/docs/data/csv/tips).
> 
> ```
> id	date	str
> 1	2020-01-01	hello world
> ```
> 
>  ```sql
> D FROM read_csv_auto('test.csv');
> ┌───────┬────────────┬─────────────┐
> │  id   │    date    │     str     │
> │ int64 │    date    │   varchar   │
> ├───────┼────────────┼─────────────┤
> │     1 │ 2020-01-01 │ hello world │
> └───────┴────────────┴─────────────┘
> D FROM read_csv_auto('test.csv', types={'date': 'VARCHAR'});
> ┌───────┬────────────┬─────────────┐
> │  id   │    date    │     str     │
> │ int64 │  varchar   │   varchar   │
> ├───────┼────────────┼─────────────┤
> │     1 │ 2020-01-01 │ hello world │
> └───────┴────────────┴─────────────┘
> ```

Perfect! I had not seen that page. Perhaps a link to that documentation belongs [here](https://duckdb.org/docs/data/csv/overview.html) (or maybe the content should be consolidated there).

**Reply by iangow** [reply]
> Perfect! I had not seen that page. Perhaps a link to that documentation belongs [here](https://duckdb.org/docs/data/csv/overview.html) (or maybe the content should be consolidated there).

Sorry. I now see that a link is there ... it's just inconspicuously placed at the bottom. Maybe need to list arguments (parameters) available for `read_csv_auto()`.

**Comment by krlmlr** [comment]
I don't see problems with adding readr as an opt-in soft dependency, and use `readr::spec_csv2()` and friends to learn about the structure of a CSV file. But perhaps this can also be solved with documentation, and some code that generates/runs the correct `read_csv_auto()` call required for a particular CSV file?

**Comment by hannes** [comment]
the duckdb_read_csv is provided for backwards compatibility and probably obsolete now. We should consider directly wrapping DuckDB's (powerful) CSV reader. Users who wish to use readr can do so and insert/query the resulting data frame directly.

**Comment by krlmlr** [comment]
The new CSV logic is powerful indeed!

Is there a way, for any given CSV file, to autodetect its format via `read_csv_auto()` and to emit the corresponding explicit `read_csv()` call?
